File: THREE/Math/Plane.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace THREE
{
    /// <summary>
    /// @author bhouston / http://clara.io
    /// @author tengge / https://github.com/tengge1
    /// </summary>
    public class Plane
    {
        public Vector3 normal;
        public double constant;

        public Plane(Vector3 normal = null, double constant = 0.0)
        {
            // normal is assumed to be normalized

            if (normal == null)
            {
                normal = new Vector3(1, 0, 0);
            }
        }

        public Plane Set(Vector3 normal, double constant)
        {
            this.normal.Copy(normal);
            this.constant = constant;

            return this;
        }

        public Plane SetComponents(double x, double y, double z, double w)
        {
            this.normal.Set(x, y, z);
            this.constant = w;

            return this;
        }

        public Plane SetFromNormalAndCoplanarPoint(Vector3 normal, Vector3 point)
        {
            this.normal.Copy(normal);
            this.constant = -point.Dot(this.normal);

            return this;
        }

        public Plane SetFromCoplanarPoints(Vector3 a, Vector3 b, Vector3 c)
        {
            var v1 = new Vector3();
            var v2 = new Vector3();

            var normal = v1.SubVectors(c, b).Cross(v2.SubVectors(a, b)).Normalize();

            // Q: should an error be thrown if normal is zero (e.g. degenerate plane)?

            this.SetFromNormalAndCoplanarPoint(normal, a);

            return this;
        }

        public Plane Clone()
        {
            return new Plane().Copy(this);
        }

        public Plane Copy(Plane plane)
        {
            this.normal.Copy(plane.normal);
            this.constant = plane.constant;

            return this;
        }

        public Plane Normalize()
        {
            // Note: will lead to a divide by zero if the plane is invalid.

            var inverseNormalLength = 1.0 / this.normal.Length();
            this.normal.MultiplyScalar(inverseNormalLength);
            this.constant *= inverseNormalLength;

            return this;
        }

        public Plane Negate()
        {
            this.constant *= -1;
            this.normal.Negate();

            return this;
        }

        public double DistanceToPoint(Vector3 point)
        {
            return this.normal.Dot(point) + this.constant;
        }

        public double DistanceToSphere(Sphere sphere)
        {
            return this.DistanceToPoint(sphere.center) - sphere.radius;
        }

        public Vector3 ProjectPoint(Vector3 point, Vector3 target = null)
        {
            if (target == null)
            {
                Console.WriteLine("THREE.Plane: .projectPoint() target is now required");
                target = new Vector3();
            }

            return target.Copy(this.normal).MultiplyScalar(-this.DistanceToPoint(point)).Add(point);
        }

        public Vector3 IntersectLine(Line3 line, Vector3 target = null)
        {
            var v1 = new Vector3();
            if (target == null)
            {
                Console.WriteLine("THREE.Plane: .intersectLine() target is now required");
                target = new Vector3();
            }

            var direction = line.Delta(v1);

            var denominator = this.normal.Dot(direction);

            if (denominator == 0)
            {
                // line is coplanar, return origin
                if (this.DistanceToPoint(line.start) == 0)
                {
                    return target.Copy(line.start);
                }

                // Unsure if this is the correct method to handle this case.
                return null;
            }

            var t = -(line.start.Dot(this.normal) + this.constant) / denominator;

            if (t < 0 || t > 1)
            {
                return null;
            }

            return target.Copy(direction).MultiplyScalar(t).Add(line.start);
        }

        public bool IntersectsLine(Line3 line)
        {
            // Note: this tests if a line intersects the plane, not whether it (or its end-points) are coplanar with it.

            var startSign = this.DistanceToPoint(line.start);
            var endSign = this.DistanceToPoint(line.end);

            return startSign < 0 && endSign > 0 || endSign < 0 && startSign > 0;
        }

        public bool IntersectsBox(Box3 box)
        {
            return box.IntersectsPlane(this);
        }

        public bool IntersectsSphere(Sphere sphere)
        {
            return sphere.IntersectsPlane(this);
        }

        public Vector3 CoplanarPoint(Vector3 target = null)
        {
            if (target == null)
            {
                Console.WriteLine("THREE.Plane: .coplanarPoint() target is now required");
                target = new Vector3();
            }

            return target.Copy(this.normal).MultiplyScalar(-this.constant);
        }

        public Plane ApplyMatrix4(Matrix4 matrix, Matrix3 optionalNormalMatrix = null)
        {

            var v1 = new Vector3();
            var m1 = new Matrix3();

            var normalMatrix = optionalNormalMatrix ?? m1.GetNormalMatrix(matrix);

            var referencePoint = this.CoplanarPoint(v1).ApplyMatrix4(matrix);

            var normal = this.normal.ApplyMatrix3(normalMatrix).Normalize();

            this.constant = -referencePoint.Dot(normal);

            return this;
        }

        public Plane Translate(Vector3 offset)
        {
            this.constant -= offset.Dot(this.normal);

            return this;
        }

        public bool Equals(Plane plane)
        {
            return plane.normal.Equals(this.normal) && (plane.constant == this.constant);
        }
    }
}

File: server/three/plane.go
// Copyright 2017-2020 The ShadowEditor Authors. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.
//
// For more information, please visit: https://github.com/tengge1/ShadowEditor
// You can also visit: https://gitee.com/tengge1/ShadowEditor
//
// This package is translated from three.js, visit `https://github.com/mrdoob/three.js`
// for more information.

package three

var _vector1 = Vector3{}
var _vector2 = Vector3{}
var _normalMatrix = Matrix3{}

// NewPlane :
func NewPlane(normal Vector3, constant float64) *Plane {
	// normal is assumed to be normalized
	return &Plane{normal, constant}
}

// Plane :
type Plane struct {
	Normal   Vector3
	Constant float64
}

// Set :
func (p Plane) Set(normal Vector3, constant float64) *Plane {
	p.Normal.Copy(normal)
	p.Constant = constant
	return &p
}

// SetComponents :
func (p Plane) SetComponents(x, y, z, w float64) *Plane {
	p.Normal.Set(x, y, z)
	p.Constant = w
	return &p
}

// SetFromNormalAndCoplanarPoint :
func (p Plane) SetFromNormalAndCoplanarPoint(normal, point Vector3) *Plane {
	p.Normal.Copy(normal)
	p.Constant = -point.Dot(p.Normal)
	return &p
}

// SetFromCoplanarPoints :
func (p Plane) SetFromCoplanarPoints(a, b, c Vector3) *Plane {
	normal := _vector1.SubVectors(c, b).Cross(*_vector2.SubVectors(a, b)).Normalize()
	// Q: should an error be thrown if normal is zero (e.g. degenerate plane)?
	p.SetFromNormalAndCoplanarPoint(*normal, a)
	return &p
}

// Clone :
func (p Plane) Clone() *Plane {
	return NewPlane(p.Normal, p.Constant).Copy(p)
}

// Copy :
func (p Plane) Copy(plane Plane) *Plane {
	p.Normal.Copy(plane.Normal)
	p.Constant = plane.Constant
	return &p
}

// Normalize :
func (p Plane) Normalize() *Plane {
	// Note: will lead to a divide by zero if the plane is invalid.
	inverseNormalLength := 1.0 / p.Normal.Length()
	p.Normal.MultiplyScalar(inverseNormalLength)
	p.Constant *= inverseNormalLength
	return &p
}

// Negate :
func (p Plane) Negate() *Plane {
	p.Constant *= -1
	p.Normal.Negate()
	return &p
}

// DistanceToPoint :
func (p Plane) DistanceToPoint(point Vector3) float64 {
	return p.Normal.Dot(point) + p.Constant
}

// DistanceToSphere :
func (p Plane) DistanceToSphere(sphere Sphere) float64 {
	return p.DistanceToPoint(sphere.Center) - sphere.Radius
}

// ProjectPoint :
func (p Plane) ProjectPoint(point, target Vector3) *Vector3 {
	return target.Copy(p.Normal).MultiplyScalar(-p.DistanceToPoint(point)).Add(point)
}

// IntersectLine :
func (p Plane) IntersectLine(line Line3, target Vector3) *Vector3 {
	direction := line.Delta(_vector1)
	denominator := p.Normal.Dot(*direction)
	if denominator == 0 {
		// line is coplanar, return origin
		if p.DistanceToPoint(line.Start) == 0 {
			return target.Copy(line.Start)
		}
		// Unsure if p is the correct method to handle p case.
		return nil
	}

	t := -(line.Start.Dot(p.Normal) + p.Constant) / denominator
	if t < 0 || t > 1 {
		return nil
	}

	return target.Copy(*direction).MultiplyScalar(t).Add(line.Start)
}

// IntersectsLine :
func (p Plane) IntersectsLine(line Line3) bool {
	// Note: p tests if a line intersects the plane, not whether it (or its end-points) are coplanar with it.
	startSign := p.DistanceToPoint(line.Start)
	endSign := p.DistanceToPoint(line.End)
	return startSign < 0 && endSign > 0 || endSign < 0 && startSign > 0
}

// IntersectsBox :
func (p Plane) IntersectsBox(box Box3) bool {
	return box.IntersectsPlane(p)
}

// IntersectsSphere :
func (p Plane) IntersectsSphere(sphere Sphere) bool {
	return sphere.IntersectsPlane(p)
}

// CoplanarPoint :
func (p Plane) CoplanarPoint(target Vector3) *Vector3 {
	return target.Copy(p.Normal).MultiplyScalar(-p.Constant)
}

// ApplyMatrix4 :
func (p Plane) ApplyMatrix4(matrix Matrix4) *Plane {
	normalMatrix := _normalMatrix.GetNormalMatrix(matrix)
	referencePoint := p.CoplanarPoint(_vector1).ApplyMatrix4(matrix)
	normal := p.Normal.ApplyMatrix3(*normalMatrix).Normalize()
	p.Constant = -referencePoint.Dot(*normal)
	return &p
}

// Translate :
func (p Plane) Translate(offset Vector3) *Plane {
	p.Constant -= offset.Dot(p.Normal)
	return &p
}

// Equals :
func (p Plane) Equals(plane Plane) bool {
	return plane.Normal.Equals(p.Normal) && plane.Constant == p.Constant
}
